wasm_memsafety_rtlib.c

```diff
@@ -64,66 +70,183 @@ static void __wasm_memsafety_table_resize(HashTable *table, size_t capacity) {
-        fprintf(stderr, "Tagging memory %p, size %zu\n", mem, aligned_size);
+        fprintf(stderr, "Tagging memory %p, requested size %zu, allocated size %zu\n", mem, requested_size, aligned_size);
 
-        __wasm_memsafety_table_set(&table, mem, aligned_size);
+        __wasm_memsafety_table_set(&table, mem, requested_size, aligned_size);
     }
 
     return mem;
 }
 
+// The free() function frees the memory space pointed to by ptr,
+// which must have been returned by a previous call to malloc() or
+// related functions.  Otherwise, or if ptr has already been freed,
+// undefined behavior occurs.  If ptr is NULL, no operation is
+// performed.
 void __wasm_memsafety_free(void *ptr) {
     TableEntry entry = __wasm_memsafety_table_remove(&table, ptr);
     if (entry.key != NULL) {
-        fprintf(stderr, "Untagging memory %p, size %zu\n", entry.key, entry.size);
-        __builtin_wasm_segment_free(entry.key, entry.size);
-        void *untagged_ptr = (void *) ((size_t) entry.key & 0xF0FFFFFFFFFFFFFF);
+        fprintf(stderr, "Untagging memory %p, requested size %zu, allocated size %zu\n", entry.key, entry.requested_size, entry.allocated_size);
+        __builtin_wasm_segment_free(entry.key, entry.allocated_size);
+        void *untagged_ptr = STRIP_MTE_TAG(entry.key);
         free(untagged_ptr);
     } else {
+        // This branch should only be entered when `free(NULL)` is called, which
+        // is safe and defined in the C standard.
         free(ptr);
     }
 }
+
+// Check if multiplying a and b would overflow.
+static inline bool check_multiplication_overflow(size_t a, size_t b) {
+    // We want to check if `a * b > SIZE_MAX`, which can be mathematically
+    // transformed to `b > SIZE_MAX / a`
+    return (a != 0 && b > (SIZE_MAX / a));
+}
+
+// The calloc() function allocates memory for an array of nmemb elements of size
+// bytes each and returns a pointer to the allocated memory. The memory is set to
+// zero. If nmemb or size is 0, then calloc() returns either NULL, or a unique
+// pointer value that can later be successfully passed to free().
+void *__wasm_memsafety_calloc(size_t nmemb, size_t size) {
+    if (check_multiplication_overflow(nmemb, size)) {
+        return NULL;
+    }
+    size_t requested_size = nmemb * size;
+
+    // If we don't early exit here on requested size 0, we would allocate
+    // 16 bytes unnecessarily.
+    if (requested_size == 0) {
+        return NULL;
+    }
+
+    void *mem = __wasm_memsafety_malloc(MTE_ALIGNMENT, requested_size);
+    if (mem) {
+        // We only set the requested number of bytes to 0, not any extra bytes
+        // that malloc might have allocated due to alignment.
+        memset(mem, 0, requested_size);
+    }
+
+    return mem;
+}
+
+static inline size_t min(size_t a, size_t b) {
+    return a < b ? a : b;
+}
+
+// TODO: what should happen if we call realloc on a ptr that was previously called with aligned_alloc? Do we have to follow that original alignment?
```

Comment:
Just glancing over the C standard, it doesn't mention alignment so I think we're good.